### PR TITLE
Rethinking 2 end of chapter solutions

### PR DESCRIPTION
**Pull request by AidanMar**
Hi there,

I've been working my way through the 2nd edition of Statistical Rethinking using pymc3. I noticed you don't have end of chapter solutions yet like you do for the 1st version. I have already produced solutions for quite a few of the chapters. They were originally intended for personal/internal use at my research group, but I thought they might be useful here. They aren't totally polished in terms of code quality or in the main text that goes with it, but I'm more than happy to take the time to bring them up to a standard that you might find satisfying. Let me know what you think. 

Thanks,
Aidan

### CODE REVIEW

File: Rethinking_2/End_of_chapter_problems/Chapter 2 - SMALL WORLDS AND LARGE WORLDS.ipynb [hunk 1]
```diff
@@ -0,0 +1,839 @@
+{
```
Comment (aloctavodia) [inline Rethinking_2/End_of_chapter_problems/Chapter 2 - SMALL WORLDS AND LARGE WORLDS.ipynb:1]: <p>Remove unused imports (I think pandas and seaborn). Reorder imports see other notebooks as example.</p><p> </p><p>Do this for all notebooks</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/resources/pull/119/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (aloctavodia) [inline Rethinking_2/End_of_chapter_problems/Chapter 2 - SMALL WORLDS AND LARGE WORLDS.ipynb:1]: <p>Add a cell above this with</p><p> </p><pre class="ql-syntax" spellcheck="false">%config Inline.figure_format = 'retina'
az.style.use('arviz-darkgrid')
az.rcParams['stats.hdi_prob'] = 0.89  
</pre><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/resources/pull/119/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (aloctavodia) [inline Rethinking_2/End_of_chapter_problems/Chapter 2 - SMALL WORLDS AND LARGE WORLDS.ipynb:1]: <p>remove this</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/resources/pull/119/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (aloctavodia) [inline Rethinking_2/End_of_chapter_problems/Chapter 2 - SMALL WORLDS AND LARGE WORLDS.ipynb:1]: <p>Remove this section</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/resources/pull/119/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (aloctavodia) [inline Rethinking_2/End_of_chapter_problems/Chapter 2 - SMALL WORLDS AND LARGE WORLDS.ipynb:1]: <p>remove this section</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/resources/pull/119/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

Comment (aloctavodia) [inline Rethinking_2/End_of_chapter_problems/Chapter 2 - SMALL WORLDS AND LARGE WORLDS.ipynb:1]: <p>no need to use <code>plt.show()</code> inside a Jupyter notebook.</p><p> </p><p>Paramter --&gt; Parameter</p><p>Prosterior --&gt; Posterior</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/resources/pull/119/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='21'/>

File: Rethinking_2/End_of_chapter_problems/Chapter 3 - SAMPLING THE IMAGINARY.ipynb [hunk 2]
```diff
@@ -0,0 +1,1360 @@
+{
```
Comment (aloctavodia) [inline Rethinking_2/End_of_chapter_problems/Chapter 3 - SAMPLING THE IMAGINARY.ipynb:1]: <p>Do not use <code>start = pm.find_MAP()</code>, in general is not a good idea. For most simple model you will not need to do burnin. Also notice that PyMC3 by defaults runs 1000 steps of tuning and those steps are discarded.</p><p> </p><p>Use ArviZ function explicetiely, like az.plot_trace, az.plot_forest (notice that the correct argument name is <code>var_names</code>)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/resources/pull/119/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (aloctavodia) [inline Rethinking_2/End_of_chapter_problems/Chapter 3 - SAMPLING THE IMAGINARY.ipynb:1]: <p>use <code>az.hdi</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/resources/pull/119/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='38'/>

Comment (aloctavodia) [inline Rethinking_2/End_of_chapter_problems/Chapter 3 - SAMPLING THE IMAGINARY.ipynb:1]: <p>add a semicolon at the end of the last line like <code>plt.legend(loc=(1.1,0.5));</code></p><p> </p><p>This supress the matplotlib output (the <code>&lt;matplotlib.legend.Legend at 0x7fea4033c3d0&gt;</code>  )in this example</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/resources/pull/119/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='59'/>

**Comment by aloctavodia** [comment]
Hi @AidanMar Thanks for your contribution I made a few comments, most of them about style. Notice that most of these comments apply to other cells and even other notebooks.

**Comment by AlexAndorra** [comment]
Thanks a lot for your contribution @AidanMar, this is really great 🤩 
To help you with @aloctavodia's suggestions, you can take a look at the [style guide](https://github.com/pymc-devs/pymc3/wiki/PyMC's-Jupyter-Notebook-Style/3a92c230b9dc8896d6c82a32f4ce25f8d9e73191) we use on this repo. You can even automate most of this, as we do on the main PyMC repo, by using `isort` and `black` via the [`nbqa` package](https://github.com/nbQA-dev/nbQA) (cc @MarcoGorelli) -- here is the [PyMC style guide](https://github.com/pymc-devs/pymc3/wiki/PyMC's-Jupyter-Notebook-Style) for more details.

**Comment by AidanMar** [comment]
Thanks both. The feedback above is appreciated. It'll probably be the weekend before I can really get to work on doing these edits, just so you don't think that I haven't given up if you don't see any activity for a few days. Looking forward to polishing them up.

**Comment by AlexAndorra** [comment]
Absolutely no problem -- we also have a lot on our plates right now and can't be as involved on this repo as we'd like, so we clearly understand 😅

**Comment by AlexAndorra** [comment]
Hey @AidanMar ! Do you think you'll get some time to address @aloctavodia's quick comments? 
We'll try to be more responsive in merging and reviewing PRs now 😉 Plus, Richard started teaching his 2020 course and recommended PyMC3 🎉 So I'm guessing this repo will be useful to quite a lot of people!

**Comment by AidanMar** [comment]
Hey @AlexAndorra. Apologies for the delays. I've had my time really be absorbed by a lot of things in the last few weeks, but I'm finally back onto this. I'll be pushing with edits and new material this weekend. Thanks for the update about his course being up and running. I'll be committing more time here now!

**Comment by AidanMar** [comment]
@AlexAndorra I'm working my way through the changes. I made edits for chapters 2 through to 4. I've made preliminary style edits for the other chapters, but I still need to reread them and make more major updates to them. I wanted to check that I'm getting everything correct with chapters 2-4 before I invest more time into the remaining chapters. Let me know what you think. All the best :) I should also have a couple more chapters from later in the book coming in soon!

**Review by AlexAndorra — CHANGES_REQUESTED** [review]
Amazing, thanks @AidanMar !
This is starting to look good 🤩 I have two main comments:

1. Instead of using the `hdi_prob` kwarg each time in `az.hdi` or `az.plot_hdi`, just set the default credible interval to 89% at the beginning of the notebooks, just after you're setting the ArviZ plotting style. You just need to do `az.rcParams['stats.hdi_prob'] = 0.89` 😉 

2. I would use Latex formatting more intensively, to better differentiate the different exercises. It'll be easier to parse through for readers, especially if they are looking for the answer to a specific exercise. You can take a look at how we formatted practice problems for the first edition to get an idea (`Rethinking` directory, on this same repository).

Thanks again for your great work, feel free to ask if you have any questions, and let me know when I can review chapters 5 through 9 then 🖖 

PS: starting from chapter 10, let's do another PR, otherwise this one will become huge. Plus, it means people will already be able to use the first 9 chapters 😉 You can do one PR per chapter for instance 🤷‍♂️

**Comment by AidanMar** [comment]
I just uploaded more changes for the remaining chapters. I'm having a problem right at the end of chapter 9 notebook. The question requires a bounded normal distribution and for that model, sampling from the posterior doesn't seem to work very well. The trace plot looks pretty crazy. Let me know if you have any pointers on why that might be happening, and how I could fix it.

Let me know what you think.

Thanks, :)

**Review by AlexAndorra — CHANGES_REQUESTED** [review]
Thanks @AidanMar !
- There are still a few changes to make, that I commented below. I did a thorough review of NBs up until chapter 6, and the needed changes are very similar in all NBs. So, I won't repeat them for chapters 7, 8 and 9, but you can already go ahead and change this in all NBs
- I'll wait for these changes to review 7 to 9 but I'll take a look soon at the problematic model you're mentioning, at the end of chapter 9 😉 
- Pre-commit checks are failing because you didn't apply `nbqa` to the notebooks. [See the NB style](https://github.com/pymc-devs/pymc3/wiki/PyMC's-Jupyter-Notebook-Style) for a quick and easy way to do this

Thanks again and feel free to ask any questions 🖖

**Comment by AidanMar** [comment]
Hi again. I think I've addressed most of the changes you recommended in the earlier chapters, and have tried to carry those through to the later chapters as far as possible. I'm surprised the pre-commit checks failed, as I installed pre-commit ran the relevant nbqa commands which allowed them to pass on my end after I made the relevant changes. Is there a specific sequence that nbqa-isort,nbqa-black and pyupgrade need to be run in order to get success on the pre-commit checks on the uploads?

**Comment by AlexAndorra** [comment]
Thanks @AidanMar, I'll review ASAP!

> Is there a specific sequence that nbqa-isort,nbqa-black and pyupgrade need to be run in order to get success on the pre-commit checks on the uploads?

Not that I know of. Did you do `pre-commit run --files notebook1.ipynb notebook2.ipynb`, as explained in the style guide?
This should make all the changes automatically and add a commit locally, that you can then push. If this command passes locally, then the pre-commit checks in CI will very probably pass here too

**Review by AlexAndorra — COMMENTED** [review]
Thanks @AidanMar, this is looking good! There are still small bits to improve, but this PR is already huge, so let's focus on getting the pre-commit checks to pass and then we'll merge 🎉 

The improvements can be done in subsequent PRs if you get some time. In particular:

- I think you don't need to cast to `pm.intX` now for categorical data (e.g in  `pm.intX(pm.Data("cid", Complete.cont_africa))` because `pm.Data` is able to do it automatically now, IIRC

- I still think you're making your life more complicated than it needs to be when dealing with `InferenceData` (I know, it's yet another stuff to learn 😅 But you'll see it's really powerful). For instance, things like:
```python
model2_post.posterior["alpha"].values.reshape(1, -1) + M * model2_post.posterior["beta_M"].values.reshape(1, -1) + np.multiply(N, model2_post.posterior["beta_N"].values.reshape(1, -1))
```
are convoluted because you're not actually using `xarray` magic. I unfortunately don't have time right now to push to this PR directly to show you an example, but I'm betting this should simply reduce to:
```python
model2_post.posterior["alpha"] + M * model2_post.posterior["beta_M"] + N * model2_post.posterior["beta_N"]
```
thanks to `xarray` magic 🤩 In other words, your shape-handling headaches are mostly gone (provided the `dims` and `coords` are correctly specified)!

Again, I think we can merge as is, but I'll [refer you back to the resources](https://github.com/pymc-devs/resources/pull/119#issuecomment-749752728) I linked too in a previous comment, and feel free to open a PR making use of `xarray` when you get the hang of it, that's a good exercise and will prove very useful to you 😉 

Hope this helps, and thanks again for your big contribution 👏

**Comment by MarcoGorelli** [comment]
Wow, this is a lot of files :)

For the checks, it may be easier to run
```
pre-commit run --from-ref=upstream/master --to-ref=HEAD --all-files
```
, assuming your `upstream` remote points to `pymc-devs/resources`

Or even just
```
pre-commit run --all-files
```
probably won't take too long here

Any issues with that, please do ping me and I'll take a look

**Comment by AidanMar** [comment]
I think I succeeded in getting the pre-commit checks to work. Hopefully able to do more revisions for you guys in near future. Looking to possibly add other chapters as well :). I know there's still quite a bit more to be done, but hopefully it's enough to start for the 2nd edition. Let me know if you have anything you think can/should be done better.

The first thing I'll focus making better use of xarray!

**Review by AlexAndorra — APPROVED** [review]
Awesome, thanks and well done on this big contribution @AidanMar 👏 
Really looking forward to your next PRs then 🥳 

> The first thing I'll focus on is making better use of xarray!

I think you'll indeed find it's a good use of your time 😉

**Pull request merged**